### PR TITLE
[Webserver] update to version 3.13.3

### PR DESCRIPTION
**Pull request by ThomasBott**
Includes all fixed and Improvements discussed in the EventGhost forum the last months. These are:
3.13.2 by david-mark 2016-11-18 20:00 UTC+1
     - bugfix (Fixed incorrect "latin1" decoding)
3.13.1 by Sem;colon 2016-11-06 01:00 UTC+1
     - bugfix (Added "self.protocol_version = 'HTTP/1.1'" for ws connections)
3.13 by Sem;colon 2016-10-29 01:00 UTC+1
     - bugfix (Send event to another webserver was never doing GET but always POST)
     - Changes to support EG v0.5 (Synchronised with the version on GitHub)
3.12.1 by Pako 2016-05-08 18:19 UTC+1
     - many changes - particularly regarding the WebSocket

### CODE REVIEW

**Comment by topic2k** [comment]
Hmm, the diff shows the whole file as changed!? If i do it in [my repo](https://github.com/EventGhost/EventGhost/compare/master...topic2k:webserver_update?expand=1) this doesn't happen. 
This makes seeing the changes a bit difficult (even if there has a lot been changed).

File: plugins/Webserver/__init__.py [hunk 1]
```diff
@@ -1,5492 +1,6594 @@
...
-            choices = self.plugin.text.modeClientChoice,
-            style=wx.RA_SPECIFY_ROWS
-        )
-        radioBoxModeClient.SetSelection(modeClient)
-        staticBox = wx.StaticBox(panel, -1, "")
-        tmpSizer = wx.GridBagSizer(2, 10)
-        txtLabel = wx.StaticText(panel,-1,self.plugin.text.host)
-        txtCtrl = wx.TextCtrl(panel,id3,"")
-        portLabel = wx.StaticText(panel,-1,self.plugin.text.port)
-        portCtrl = wx.TextCtr
```
Comment (topic2k) [inline plugins/Webserver/__init__.py]: Too broad exception clause: specifiy exception class (`ImportError`)

Comment (topic2k) [inline plugins/Webserver/__init__.py]: `true` is unused

Comment (topic2k) [inline plugins/Webserver/__init__.py]: `false` is unused

**Comment by per1234** [comment]
I think it would be better if each version was added as a separate commit. There are a lot of unrelated changes mashed into a single commit here which makes troubleshooting much more difficult.

**Comment by topic2k** [comment]
@ThomasBott, will you make the suggested changes? Would love to have it in the next beta release.

**Comment by ThomasBott** [comment]
@topic2k Yes, I got in contact with Pako to verify if if my proposed changes about your remarks (thx for checking the code btw.) are correct. We came up with this now:

- line 152: syspath is unused -> Yes, not needed any longer, can be removed
- line 155: strftime is unused -> No, still used (line 697)
- line 161: cpy is unused -> Yes, not needed any longer, can be removed
- line 175: Too broad exception clause: specifiy exception class (ImportError) -> Yes
- line 190: true is unused -> No, used for external scripting to be more convenient(JS uses true, Python uses True)
- line 191: false is unused -> No, used for external scripting to be more convenient(JS uses false, Python uses False)
- line 2198: unresolved reference plugin -> should be self.plugin
- line 2252: unresolved reference args -> should be None
- line 2253: unresolved reference kwargs -> should be None
- line 3497: unresolved reference plugin -> should be self.plugin
- line 3562: unresolved reference args -> should be None
- line 3563: unresolved reference kwargs -> should be None

I'll apply the fixes!

@per1234 not this time, as the version from EG 0.5 drifted apart from the one in the forum and I had to merge both. Also most of the changes were made in one release anyway.

**Comment by per1234** [comment]
>not this time, as the version from EG 0.5 drifted apart from the one in the forum and I had to merge both. Also most of the changes were made in one release anyway.

Well I disagree with that approach. I'm not running the repo so I can only make suggestions but I recommend reading the ["Keep Your Commits Atomic" article](http://www.freshconsulting.com/atomic-commits/) linked from [CONTRIBUTING.md](https://github.com/EventGhost/EventGhost/blob/master/CONTRIBUTING.md).

Also, what's the point of having versions that don't actually exist?

**Comment by ThomasBott** [comment]
> Also, what's the point of having versions that don't actually exist?
They do exist, you can download them from the forum.
Why should I create a pull request with a version that doesn't work with this branch?
If you like to do this, do it, you can find the files in the forum, but it doesn't make much sense to me...

**Comment by per1234** [comment]
> Why should I create a pull request with a version that doesn't work with this branch?

The only version that's not compatible with EventGhost 0.5.x is 3.12.1. 3.13.0 and newer are compatible so there's no good reason for not putting them in separate commits. I still think 3.12.1 should be a separate commit with a commit message that clearly states the incompatibility. Even though it's not compatible it's still important for the sake of the diff.
> If you like to do this, do it, you can find the files in the forum, but it doesn't make much sense to me...

I'm happy to do so if topic2k supports it but I'd rather that this pull request was made better than to create a separate one. Also, I think it's an important discussion to have about following the guidelines of CONTRIBUTING.md to maintain a sane commit history. It may seem uptight but it's really important when you have many people working on a long term project. When you are trying to track down a bug or figure out why a change was made you realize that (good) commits really do count!

**Comment by ThomasBott** [comment]
> The only version that's not compatible with EventGhost 0.5.x is 3.12.1. 3.13.0 and newer are compatible so there's no good reason for not putting them in separate commits.

You are right. To be honest I'm not very familiar with GitHub, I was happy that I (more or less) successfully created a Pull Request, The last thing I was thinking of was "well, but first I need to make some commits for the old versions". I don't know if it is possible to create commits for previous versions afterwards or how this would work... But I'll not discard this Pull Request and create a new one just do have the diff of two one-line changes. Hell, we were developing EG without a diff for so many years... The big change was in 3.12.1, and GitHub can't diff this one anyway... It would be unnecessarily confusing to make a commit for this (it was actually released before the version that has been integrated in EG 0.5, as the one in EG 0.5 has been modified without increasing the version), you may as well create commits for the 100 versions prior to that one, if you like to see the compete history. 

I do understand the purpose or the atomic principle and I agree with it. I will follow it thoroughly in the future, if all development is done on GitHub.
Currently we are in a migration phase – some code is developed in the forum, some is developed on GitHub. But since no one likes to maintain their code on two places, one will always suffer, until all development is moved. (and I’m taking about the people here, not the code)

**Comment by per1234** [comment]
>I'm not very familiar with GitHub

There's a bit of a learning curve for git and GitHub but once you get a bit comfortable it's amazing, both for collaboration and source control of your personal projects.
> I don't know if it is possible to create commits for previous versions afterwards or how this would work.

It is possible, but you have to actually use git rather than the simplified interface that GitHub or GitHub Desktop provides. Any commit you push to your master branch will be included in this pull request automatically. Since you would need to change the commit history of the branch you need to do a force push(git push -f). Generally changing history is considered a bad thing but it's fine for a development branch, you just don't want to ever do it to a branch where other people are counting on the history not changing. If you need more details on how to split the changes into multiple commits I'm happy to help.
>Hell, we were developing EG without a diff for so many years... 

Not true. You can see the full commit history of the plugin back to 2011 here:
https://github.com/EventGhost/EventGhost/commits/master/plugins/Webserver/__init__.py
The previous way of developing seems to have been that people would post code to the forum and then someone like Pako would then merge it into the repository so you may not have been aware of that part of things.
>you may as well create commits for the 100 versions prior to that one, if you like to see the compete history.

As I just showed, they are already there. All the commits were migrated from the SourceForge repository.
>the one in EG 0.5 has been modified without increasing the version

The version number doesn't necessarily need to be increased on every change, only at each release, which should actually be a separate commit just for the version bump(following semver). You can use the commit hash to identify any point in the history of the code.
>Currently we are in a migration phase – some code is developed in the forum, some is developed on GitHub.

You can develop wherever you like but it's absolutely essential that people are pulling the latest source from this repository and pushing any changes back here in an atomic manner. There's no possible compromise on that. If that's not happening then it's a very serious issue. If this is something people aren't aware of then someone needs to make a post on the forum and also a note should be added to to contributing.rst.

**Comment by ThomasBott** [comment]
@per1234 Thank you for all the information, I really appreciate!
Unfortunately, I don't have time at the moment to dig deeper in the git topic, so I'll keep maintaining the forum thread for now and leave the GitHub commits to someone that's more familiar with how it works.
I'll get back here once I figured it out ;-)

**Pull request closed without merging**